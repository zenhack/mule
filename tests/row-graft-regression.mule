#XFAIL: type formatting
#
# This is a regression test for a bug where we weren't grafting correctly;
# the type should be:
#
#   all a b r. (a -> b) -> {z: a, ...r} -> {y: b, z: a, ...r}
#
# but with the bug it was:
#
#   all a b c r. (a -> b) -> {z: c, ...r} -> {y: b, z: c, ...r}
fn f arg. arg where { y = f arg.z }
