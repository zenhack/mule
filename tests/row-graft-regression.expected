
Desugared:

	(fn f (fn arg (((.= y) arg) (f ((. z) arg)))))

inferred type:

	(all a
	 (all b
	  ((a -> b) ->
	   (all c
	    (all d
	     ((record (types (...d)) (values ((z a) ...c))) ->
	      (record (types (...d)) (values ((y b) (z a) ...c)))))))))

Runtime term:

	(Lam
	 (0 ()
	  (Lam
	   (1 ()
	    (App
	     ((App
	       ((Lam
	         (0 () (Lam (1 () (Update (old (Var 1)) (label y) (field (Var 0)))))))
	        (Var 0)))
	      (App ((Var 1) (App ((GetField z) (Var 0)))))))))))

Evaluated:

	(Lam
	 (0 ()
	  (Lam
	   (1 ()
	    (App
	     ((App
	       ((Lam
	         (0 () (Lam (1 () (Update (old (Var 1)) (label y) (field (Var 0)))))))
	        (Var 0)))
	      (App ((Var 1) (App ((GetField z) (Var 0)))))))))))
